File: TS/系统/03．技能系统/00．技能模板+函数/04．机制组件/05．机制单位/03．固定受击次数机制单位.ts
/** @noSelfInFile */

import {
  创建可攻击机制单位,
  type 可攻击机制单位参数,
  type 可攻击机制单位实例,
} from "./01．可攻击机制单位";

const jass = require("jass.common") as any;
const japi = require("jass.japi") as any;
const GetUnitStateJapi = japi.GetUnitState as (this: void, unit: any, state: any) => number;

const GetHandleId = jass.GetHandleId as (h: any) => number;
const GetUnitState = jass.GetUnitState as (unit: any, state: any) => number;
const SetUnitState = jass.SetUnitState as (unit: any, state: any, value: number) => void;
const UNIT_STATE_LIFE = jass.UNIT_STATE_LIFE as any;
const UNIT_STATE_MAX_LIFE = jass.UNIT_STATE_MAX_LIFE as any;

const { registerDamageModifier } = require("系统.04．伤害系统.00．伤害计算.06．伤害修正回调") as {
  registerDamageModifier: (this: void, callback: (this: void, context: any) => number, priority?: number) => number;
};

export type 固定受击次数计数模式 = "任意伤害" | "纯普攻" | "最终伤害阈值" | "纯普攻或最终伤害阈值";

export interface 固定受击次数机制单位参数 extends 可攻击机制单位参数 {
  受击次数: number;
  每次伤害扣除次数?: number;
  计数模式?: 固定受击次数计数模式;
  最终伤害计数阈值?: number;
  未计数伤害无效?: boolean;
  同步生命条?: boolean;
  过滤伤害?: (this: void, context: any) => boolean;
  on受击?: (this: void, 单位: any, 剩余次数: number, context: any) => void;
  on击破?: (this: void, 单位: any, context: any) => void;
}

export interface 固定受击次数机制单位实例 extends 可攻击机制单位实例 {
  读取剩余次数(): number;
  设置剩余次数(次数: number): void;
}

interface 固定受击次数记录 {
  实例: 固定受击次数机制单位实例实现;
  参数: 固定受击次数机制单位参数;
}

const 固定受击次数单位表: Record<number, 固定受击次数记录 | undefined> = {};
let 固定受击次数伤害修正已注册 = false;

function 取单位ID(this: void, 单位: any): number {
  if (单位 == null || 单位 === 0) return 0;
  return GetHandleId(单位) || 0;
}

function 规整次数(this: void, 次数: number): number {
  if (次数 == null || 次数 !== 次数 || 次数 < 0) return 0;
  return math.floor(次数);
}

function 最终伤害达到计数阈值(this: void, 参数: 固定受击次数机制单位参数, context: any): boolean {
  const 阈值 = 参数.最终伤害计数阈值 ?? 0;
  return context.currentDamage > 阈值;
}

function 本次伤害是否计数(this: void, 参数: 固定受击次数机制单位参数, context: any): boolean {
  const 模式 = 参数.计数模式 ?? "任意伤害";
  if (模式 === "任意伤害") return true;
  const 是纯普攻 = context.isNormalAttack === true && context.isSkillAttack !== true && context.isSkillDamage !== true;
  if (模式 === "纯普攻") return 是纯普攻;
  const 达到阈值 = 最终伤害达到计数阈值(参数, context);
  if (模式 === "最终伤害阈值") return 达到阈值;
  return 是纯普攻 || 达到阈值;
}

function 固定受击次数伤害修正(this: void, context: any): number {
  const id = 取单位ID(context.target);
  if (id === 0) return context.currentDamage;
  const 记录 = 固定受击次数单位表[id];
  if (记录 == null) return context.currentDamage;
  if (context.currentDamage <= 0) return context.currentDamage;
  if (记录.参数.过滤伤害 != null && !记录.参数.过滤伤害(context)) return context.currentDamage;
  if (!本次伤害是否计数(记录.参数, context)) {
    return 记录.参数.未计数伤害无效 === true ? 0 : context.currentDamage;
  }

  const 扣除次数 = 规整次数(记录.参数.每次伤害扣除次数 ?? 1);
  if (扣除次数 <= 0) return 0;

  记录.实例.设置剩余次数(记录.实例.读取剩余次数() - 扣除次数);
  const 剩余次数 = 记录.实例.读取剩余次数();
  if (记录.参数.同步生命条 === true && 剩余次数 > 0) {
    const 总次数 = 规整次数(记录.参数.受击次数);
    if (总次数 > 0) {
      SetUnitState(记录.实例.单位, UNIT_STATE_LIFE, GetUnitStateJapi(记录.实例.单位, UNIT_STATE_MAX_LIFE) * 剩余次数 / 总次数);
    }
  }
  if (记录.参数.on受击 != null) 记录.参数.on受击(记录.实例.单位, 剩余次数, context);

  if (剩余次数 <= 0) {
    delete 固定受击次数单位表[id];
    if (记录.参数.on击破 != null) 记录.参数.on击破(记录.实例.单位, context);
    记录.实例.销毁();
  }
  return 0;
}

function 确保固定受击次数伤害修正(this: void): void {
  if (固定受击次数伤害修正已注册) return;
  固定受击次数伤害修正已注册 = true;
  registerDamageModifier(固定受击次数伤害修正, 120);
}

class 固定受击次数机制单位实例实现 implements 固定受击次数机制单位实例 {
  readonly 单位: any;
  readonly ID: number;
  private 基础实例: 可攻击机制单位实例;
  private 剩余次数: number;
  private 已销毁 = false;

  constructor(基础实例: 可攻击机制单位实例, 剩余次数: number) {
    this.基础实例 = 基础实例;
    this.单位 = 基础实例.单位;
    this.ID = 基础实例.ID;
    this.剩余次数 = 剩余次数;
  }

  是否存活(): boolean {
    return !this.已销毁 && this.剩余次数 > 0 && this.基础实例.是否存活();
  }

  读取剩余次数(): number {
    return this.剩余次数;
  }

  设置剩余次数(次数: number): void {
    this.剩余次数 = 规整次数(次数);
  }

  销毁(): void {
    if (this.已销毁) return;
    this.已销毁 = true;
    delete 固定受击次数单位表[this.ID];
    this.基础实例.销毁();
  }
}

export function 创建固定受击次数机制单位(this: void, 参数: 固定受击次数机制单位参数): 固定受击次数机制单位实例 | undefined {
  确保固定受击次数伤害修正();
  const 基础实例 = 创建可攻击机制单位({
    ...参数,
    最大生命: 参数.最大生命 ?? 999999,
    生命值受小怪倍率: false,
  });
  if (基础实例 == null) return undefined;

  const 实例 = new 固定受击次数机制单位实例实现(基础实例, 规整次数(参数.受击次数));
  固定受击次数单位表[实例.ID] = { 实例, 参数 };
  return 实例;
}

File: src/系统/03．技能系统/00．技能模板+函数/04．机制组件/05．机制单位/03．固定受击次数机制单位.lua
local ____lualib = require("lualib_bundle")
local __TS__Delete = ____lualib.__TS__Delete
local __TS__Class = ____lualib.__TS__Class
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign
local __TS__New = ____lualib.__TS__New
local ____exports = {}
local ____01_FF0E_53EF_653B_51FB_673A_5236_5355_4F4D = require("系统.03．技能系统.00．技能模板+函数.04．机制组件.05．机制单位.01．可攻击机制单位")
local _____521B_5EFA_53EF_653B_51FB_673A_5236_5355_4F4D = ____01_FF0E_53EF_653B_51FB_673A_5236_5355_4F4D["创建可攻击机制单位"]
local jass = require("jass.common")
local GetHandleId = jass.GetHandleId
local ____require_result_0 = require("系统.04．伤害系统.00．伤害计算.06．伤害修正回调")
local registerDamageModifier = ____require_result_0.registerDamageModifier
local _____56FA_5B9A_53D7_51FB_6B21_6570_5355_4F4D_8868 = {}
local _____56FA_5B9A_53D7_51FB_6B21_6570_4F24_5BB3_4FEE_6B63_5DF2_6CE8_518C = false
local function _____53D6_5355_4F4DID(_____5355_4F4D)
    if _____5355_4F4D == nil or _____5355_4F4D == 0 then
        return 0
    end
    return GetHandleId(_____5355_4F4D) or 0
end
local function _____89C4_6574_6B21_6570(_____6B21_6570)
    if _____6B21_6570 == nil or _____6B21_6570 ~= _____6B21_6570 or _____6B21_6570 < 0 then
        return 0
    end
    return math.floor(_____6B21_6570)
end
local function _____56FA_5B9A_53D7_51FB_6B21_6570_4F24_5BB3_4FEE_6B63(context)
    local id = _____53D6_5355_4F4DID(context.target)
    if id == 0 then
        return context.currentDamage
    end
    local _____8BB0_5F55 = _____56FA_5B9A_53D7_51FB_6B21_6570_5355_4F4D_8868[id]
    if _____8BB0_5F55 == nil then
        return context.currentDamage
    end
    if context.currentDamage <= 0 then
        return context.currentDamage
    end
    if _____8BB0_5F55["参数"]["过滤伤害"] ~= nil and not _____8BB0_5F55["参数"]["过滤伤害"](context) then
        return context.currentDamage
    end
    local _____6263_9664_6B21_6570 = _____89C4_6574_6B21_6570(_____8BB0_5F55["参数"]["每次伤害扣除次数"] or 1)
    if _____6263_9664_6B21_6570 <= 0 then
        return 0
    end
    local ____self_2 = _____8BB0_5F55["实例"]
    local ____self_2__8BBE_7F6E_5269_4F59_6B21_6570_3 = ____self_2["设置剩余次数"]
    local ____self_1 = _____8BB0_5F55["实例"]
    ____self_2__8BBE_7F6E_5269_4F59_6B21_6570_3(
        ____self_2,
        ____self_1["读取剩余次数"](____self_1) - _____6263_9664_6B21_6570
    )
    local ____self_4 = _____8BB0_5F55["实例"]
    local _____5269_4F59_6B21_6570 = ____self_4["读取剩余次数"](____self_4)
    if _____8BB0_5F55["参数"]["on受击"] ~= nil then
        _____8BB0_5F55["参数"]["on受击"](_____8BB0_5F55["实例"]["单位"], _____5269_4F59_6B21_6570, context)
    end
    if _____5269_4F59_6B21_6570 <= 0 then
        __TS__Delete(_____56FA_5B9A_53D7_51FB_6B21_6570_5355_4F4D_8868, id)
        if _____8BB0_5F55["参数"]["on击破"] ~= nil then
            _____8BB0_5F55["参数"]["on击破"](_____8BB0_5F55["实例"]["单位"], context)
        end
        local ____self_5 = _____8BB0_5F55["实例"]
        ____self_5["销毁"](____self_5)
    end
    return 0
end
local function _____786E_4FDD_56FA_5B9A_53D7_51FB_6B21_6570_4F24_5BB3_4FEE_6B63()
    if _____56FA_5B9A_53D7_51FB_6B21_6570_4F24_5BB3_4FEE_6B63_5DF2_6CE8_518C then
        return
    end
    _____56FA_5B9A_53D7_51FB_6B21_6570_4F24_5BB3_4FEE_6B63_5DF2_6CE8_518C = true
    registerDamageModifier(_____56FA_5B9A_53D7_51FB_6B21_6570_4F24_5BB3_4FEE_6B63, 120)
end
local _____56FA_5B9A_53D7_51FB_6B21_6570_673A_5236_5355_4F4D_5B9E_4F8B_5B9E_73B0 = __TS__Class()
_____56FA_5B9A_53D7_51FB_6B21_6570_673A_5236_5355_4F4D_5B9E_4F8B_5B9E_73B0.name = "固定受击次数机制单位实例实现"
function _____56FA_5B9A_53D7_51FB_6B21_6570_673A_5236_5355_4F4D_5B9E_4F8B_5B9E_73B0.prototype.____constructor(self, _____57FA_7840_5B9E_4F8B, _____5269_4F59_6B21_6570)
    self["已销毁"] = false
    self["基础实例"] = _____57FA_7840_5B9E_4F8B
    self["单位"] = _____57FA_7840_5B9E_4F8B["单位"]
    self.ID = _____57FA_7840_5B9E_4F8B.ID
    self["剩余次数"] = _____5269_4F59_6B21_6570
end
_____56FA_5B9A_53D7_51FB_6B21_6570_673A_5236_5355_4F4D_5B9E_4F8B_5B9E_73B0.prototype["是否存活"] = function(self)
    return not self["已销毁"] and self["剩余次数"] > 0 and self["基础实例"]["是否存活"]()
end
_____56FA_5B9A_53D7_51FB_6B21_6570_673A_5236_5355_4F4D_5B9E_4F8B_5B9E_73B0.prototype["读取剩余次数"] = function(self)
    return self["剩余次数"]
end
_____56FA_5B9A_53D7_51FB_6B21_6570_673A_5236_5355_4F4D_5B9E_4F8B_5B9E_73B0.prototype["设置剩余次数"] = function(self, _____6B21_6570)
    self["剩余次数"] = _____89C4_6574_6B21_6570(_____6B21_6570)
end
_____56FA_5B9A_53D7_51FB_6B21_6570_673A_5236_5355_4F4D_5B9E_4F8B_5B9E_73B0.prototype["销毁"] = function(self)
    if self["已销毁"] then
        return
    end
    self["已销毁"] = true
    __TS__Delete(_____56FA_5B9A_53D7_51FB_6B21_6570_5355_4F4D_8868, self.ID)
    self["基础实例"]["销毁"]()
end
____exports["创建固定受击次数机制单位"] = function(_____53C2_6570)
    _____786E_4FDD_56FA_5B9A_53D7_51FB_6B21_6570_4F24_5BB3_4FEE_6B63()
    local _____57FA_7840_5B9E_4F8B = _____521B_5EFA_53EF_653B_51FB_673A_5236_5355_4F4D(__TS__ObjectAssign({}, _____53C2_6570, {["最大生命"] = _____53C2_6570["最大生命"] or 999999, ["生命值受小怪倍率"] = false}))
    if _____57FA_7840_5B9E_4F8B == nil then
        return nil
    end
    local _____5B9E_4F8B = __TS__New(
        _____56FA_5B9A_53D7_51FB_6B21_6570_673A_5236_5355_4F4D_5B9E_4F8B_5B9E_73B0,
        _____57FA_7840_5B9E_4F8B,
        _____89C4_6574_6B21_6570(_____53C2_6570["受击次数"])
    )
    _____56FA_5B9A_53D7_51FB_6B21_6570_5355_4F4D_8868[_____5B9E_4F8B.ID] = {["实例"] = _____5B9E_4F8B, ["参数"] = _____53C2_6570}
    return _____5B9E_4F8B
end
return ____exports
